Cap cookie lifetimes to 7 days for responses from third party IP addresses

whsieh opened the pull request:
#### b0305b173106ba984cbc0475b3681daea137390c
<pre>
Cap cookie lifetimes to 7 days for responses from third party IP addresses
<a href="https://bugs.webkit.org/show_bug.cgi?id=246477">https://bugs.webkit.org/show_bug.cgi?id=246477</a>
rdar://100831206

Reviewed by John Wilander.

Safari currently caps the lifetime of cookies to 7 days, if third-party CNAME cloaking is detected.
This helps to mitigate many instances where CNAME cloaking is used to store cookies on device (in
the first party context) for far longer than a third party cookie would normally be allowed to;
however, in the case where the resolved CNAME is empty, we end up skipping this mitigation
altogether.

This means that websites can use direct A/AAAA records (instead of CNAME mapping) to cloak third
party requests as first party and subsequently store cookies in the first party context, bypassing
the aforementioned defense.

To strengthen our existing protections, we implement a heuristic to fall back on comparing resolved
IP addresses only in the case where the resolved CNAME of the incoming response is empty. If the IP
address of the response is _mostly_ different than the IP address of the main resource response
(i.e. by comparing the matching subnet mask length of the two addresses), then we apply the same
level of mitigation as we otherwise would for third party CNAMEs.

For now, the minimum matching subnet mask length to consider as &quot;third party&quot; or not is arbitrarily
chosen to be half the IP address length (i.e. 16 for IPv4, and 64 for IPv6). This could be enhanced
in the future, given facilities to query for the IP network block that contains the main resource&apos;s
IP address and checking whether the incoming response address falls within that range.

* Source/WebCore/platform/network/DNS.cpp:
(WebCore::IPAddress::isolatedCopy const):

Add an `isolatedCopy` method, so that we&apos;re able to perform a cross-thread copy of `IPAddress`.

(WebCore::IPAddress::matchingNetMaskLength const):

Add a helper method to compute the length of the matching subnet mask between the current IP address
and the given address. If the two IP addresses are of different families (i.e. v4 and v6), this
method returns 0.

* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::fromSockAddrIn6):

Minor style fix - add a missing space after the initializer.

* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::shouldApplyCookiePolicyForThirdPartyCloaking const):

Adjust this to check for third party IP addresses, in the case where the incoming response&apos;s CNAME
is empty.

(WebKit::NetworkDataTaskCocoa::updateFirstPartyInfoForSession):
(WebKit::shouldCapCookieExpiryForThirdPartyIPAddress):
(WebKit::NetworkDataTaskCocoa::applyCookiePolicyForThirdPartyCloaking):
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::shouldApplyCookiePolicyForThirdPartyCNAMECloaking const): Deleted.
(WebKit::NetworkDataTaskCocoa::applyCookiePolicyForThirdPartyCNAMECloaking): Deleted.

Rename these to reference &quot;ThirdPartyCloaking&quot; instead of &quot;ThirdPartyCNAMECloaking&quot;, since this now
applies to both.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressTests.cpp: Added.
(TestWebKitAPI::TEST):

Add a couple of API tests to exercise the new functionality in `WebCore::IPAddress`.

Canonical link: <a href="https://commits.webkit.org/255849@main">https://commits.webkit.org/255849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/005e133a23727555a4e1f183885e46901e7ac172

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93759 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2954 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103403 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163724 "Found 1 new test failure: js/dom/Promise-reject-large-string.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97752 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2970 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31207 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86096 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99418 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2117 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80198 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29141 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84043 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72096 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37602 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17596 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35452 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18856 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4041 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39330 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41414 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38087 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebCore/platform/network/DNS.cpp:
```diff
@@ -76,4 +76,50 @@ std::optional<IPAddress> IPAddress::fromString(const String& string)
+    if (isIPv6())
+        return IPAddress { ipv6Address() };
+
+    ASSERT_NOT_REACHED();
```
Could we have a release assert here? I always prefer those.

whsieh replied:
Sounds good — changed to `RELEASE_ASSERT_NOT_REACHED();`.

johnwilander commented on Source/WebCore/platform/network/DNS.cpp:
```diff
@@ -76,4 +76,50 @@ std::optional<IPAddress> IPAddress::fromString(const String& string)
+        otherAddressData = reinterpret_cast<const unsigned char*>(&other.ipv6Address());
+    }
+
+    if (!addressLengthInBytes)
```
Maybe a comment on what this case means? Is it expected?

whsieh replied:
Yes, it's expected in the case where the two `IPAddress`es are of different family (e.g. comparing v4 vs. v6) — I can see how this code looks confusing though, since it's not immediately obvious that this early return exists to catch that case.

I'll make this clearer by replacing this separate early return with an `else return;` above.

johnwilander commented on Source/WebCore/platform/network/DNS.cpp:
```diff
@@ -76,4 +76,50 @@ std::optional<IPAddress> IPAddress::fromString(const String& string)
+        matchingBits += 8;
+    }
+
+    if (firstNonMatchingIndex)
```
I would like a short comment describing that this code accomplishes. I know we typically don't add comments but I think complicated logic deserves one.

whsieh replied:
Sounds good — added this comment:

```
// Count the number of matching leading bits in the first byte that is different between the two addresses.
// For instance, 0xF2 and 0xF3 would have 7 matching bits.
```

johnwilander commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -360,7 +360,11 @@ std::optional<WebCore::IPAddress> NetworkSession::firstPartyHostIPAddress(const
         return std::nullopt;
 
-    return m_firstPartyHostIPAddresses.get(firstPartyHost);
+    auto iterator = m_firstPartyHostIPAddresses.find(firstPartyHost);
```
Is this a bug fix?

whsieh replied:
In a sense! I added this (currently unused) helper in the previous patch, and then decided that returning `nullopt` in both of these cases would be cleaner than returning either `nullopt`, or the empty value, _or_ a valid `IPAddress`.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h:
```diff
@@ -101,8 +101,8 @@ class NetworkDataTaskCocoa final : public NetworkDataTask {
     void updateFirstPartyInfoForSession(const URL&);
-    bool shouldApplyCookiePolicyForThirdPartyCNAMECloaking() const;
-    void applyCookiePolicyForThirdPartyCNAMECloaking(const WebCore::ResourceRequest&);
+    bool shouldApplyCookiePolicyForThirdPartyCloaking() const;
```
I like this renaming.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -179,9 +179,25 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
+    return cappedCookies;
+}
+
+static bool shouldCapCookieExpiry(const WebCore::IPAddress& remote, const WebCore::IPAddress& firstParty)
```
I think this one is too broadly named. Or do you think the parameter types gives it away in the signature?

whsieh replied:
Indeed — the intent was for the parameter types to make the purpose clear, but I can make this more explicit by renaming it to `shouldCapCookieExpiryForThirdPartyIPAddress`.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -193,38 +209,47 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
-    auto firstPartyHostCNAME = networkSession()->firstPartyHostCNAMEDomain(firstPartyURL.host().toString());
+    auto firstPartyHostName = firstPartyURL.host().toString();
+    auto firstPartyHostCNAME = networkSession()->firstPartyHostCNAMEDomain(firstPartyHostName);
+    auto firstPartyAddress = networkSession()->firstPartyHostIPAddress(firstPartyHostName);
```
Could any perf cost be avoided by handling one case before even looking at the other case? Nah, now I see you get these to pass into the cookie transform callback.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -193,38 +209,47 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
         if (!cnameDomain.matches(firstPartyURL) && (!firstPartyHostCNAME || cnameDomain != *firstPartyHostCNAME)) {
-
-            NSUInteger count = [cookiesSetInResponse count];
-            // Don't use RetainPtr here. This array has to be retained and
```
I remember adding this comment (and changing the code accordingly) because of an issue in CFNetwork's code. Have you made sure we're not regressing that now? I think it was for the zero cookie case. Please retain the comment if it is still relevant because it took us a long time to debug.

whsieh replied:
Sounds good — I put this comment back.

I believe what the comment is _probably_ referring to is how CFNetwork doesn't immediately retain the array we hand it back in the cookie transform block; this means that if we use a `RetainPtr` here, the array will be destroyed after the function call, before CFNetwork actually uses it, which would cause a UAF.

johnwilander commented on Tools/TestWebKitAPI/Tests/WebCore/IPAddressTests.cpp:
```diff
@@ -0,0 +1,67 @@
+    auto address3 = WebCore::IPAddress::fromString("2001:db8::1234:5678"_s);
+    auto address4 = WebCore::IPAddress::fromString("2001:db8::1111:0000"_s);
+    auto address5 = WebCore::IPAddress::fromString("::1234:5678"_s);
+    auto address6 = WebCore::IPAddress::fromString("::"_s);
```
Do we know that the IP addresses we get are always formed this way? Because the IP address spec allows form some weird formats which have historically been used to circumvent IP address detectors. Maybe that won't matter here because we're failing safe, i.e. failing *with* cap?

whsieh replied:
That's a good question — from code inspection, it seems like CFNetwork calls into libnetcore, which then calls into `getnameinfo`, which ultimately uses `inet_ntop`. Since we're using the symmetric API (`inet_pton`) to parse this string back into an IP address, I believe we _should_ be safe here in terms of being able to parse the address strings that CFNetwork hands us.

johnwilander reviewed (approved):
"Safari currently caps lifetimes of CNAME-cloaked cookies to 7 days, if the resolved host CNAME is different from the first party's CNAME." This does not fully capture the existing policy since the true first party doesn't need to have a CNAME for the cap to be applied. It may be easier to just refer to "third-party CNAME cloaking" which we have defined in a table.

OK to merge but please work through my comments and make changes if appropriate.

whsieh committed:
Cap cookie lifetimes to 7 days for responses from third party IP addresses
https://bugs.webkit.org/show_bug.cgi?id=246477
rdar://100831206

Reviewed by John Wilander.

Safari currently caps the lifetime of cookies to 7 days, if third-party CNAME cloaking is detected.
This helps to mitigate many instances where CNAME cloaking is used to store cookies on device (in
the first party context) for far longer than a third party cookie would normally be allowed to;
however, in the case where the resolved CNAME is empty, we end up skipping this mitigation
altogether.

This means that websites can use direct A/AAAA records (instead of CNAME mapping) to cloak third
party requests as first party and subsequently store cookies in the first party context, bypassing
the aforementioned defense.

To strengthen our existing protections, we implement a heuristic to fall back on comparing resolved
IP addresses only in the case where the resolved CNAME of the incoming response is empty. If the IP
address of the response is _mostly_ different than the IP address of the main resource response
(i.e. by comparing the matching subnet mask length of the two addresses), then we apply the same
level of mitigation as we otherwise would for third party CNAMEs.

For now, the minimum matching subnet mask length to consider as "third party" or not is arbitrarily
chosen to be half the IP address length (i.e. 16 for IPv4, and 64 for IPv6). This could be enhanced
in the future, given facilities to query for the IP network block that contains the main resource's
IP address and checking whether the incoming response address falls within that range.

* Source/WebCore/platform/network/DNS.cpp:
(WebCore::IPAddress::isolatedCopy const):

Add an `isolatedCopy` method, so that we're able to perform a cross-thread copy of `IPAddress`.

(WebCore::IPAddress::matchingNetMaskLength const):

Add a helper method to compute the length of the matching subnet mask between the current IP address
and the given address. If the two IP addresses are of different families (i.e. v4 and v6), this
method returns 0.

* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::fromSockAddrIn6):

Minor style fix - add a missing space after the initializer.

* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::shouldApplyCookiePolicyForThirdPartyCloaking const):

Adjust this to check for third party IP addresses, in the case where the incoming response's CNAME
is empty.

(WebKit::NetworkDataTaskCocoa::updateFirstPartyInfoForSession):
(WebKit::shouldCapCookieExpiryForThirdPartyIPAddress):
(WebKit::NetworkDataTaskCocoa::applyCookiePolicyForThirdPartyCloaking):
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::shouldApplyCookiePolicyForThirdPartyCNAMECloaking const): Deleted.
(WebKit::NetworkDataTaskCocoa::applyCookiePolicyForThirdPartyCNAMECloaking): Deleted.

Rename these to reference "ThirdPartyCloaking" instead of "ThirdPartyCNAMECloaking", since this now
applies to both.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressTests.cpp: Added.
(TestWebKitAPI::TEST):

Add a couple of API tests to exercise the new functionality in `WebCore::IPAddress`.

Canonical link: https://commits.webkit.org/255849@main
